scheduler: contextual logging - WithName and WithValue

knelasevero opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

<!--
Add one of the following kinds:
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This makes kube-scheduler log output more useful.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Related-to: https://github.com/kubernetes/kubernetes/issues/91633 Built on top of https://github.com/kubernetes/kubernetes/pull/110833 (Just rebased latest changes)

#### Special notes for your reviewer:

The basic work is in the initial commit of that WIP PR. The other commits (in that WIP PR) start with full instrumentation and then gradually reduce the runtime overhead.

I chose to create this PR already pointing to the [scheduler: use WithName and WithValues](https://github.com/kubernetes/kubernetes/pull/110833/commits/01b8046c58e288a0a9d171235fb37d4abe0acc3e) commit. 

I ran the perf tests a few times, and yeah, [we have some overhead](https://docs.google.com/spreadsheets/d/1lqUDnx6WRGB5Uu1-Xgaa_sG1bQsINA_SZ2fkay38gWs/edit#gid=0), but I personally think it is worth it [?]. Whole idea was to have named loggers initially. I think this is a nice first step and we can go to the other more specific improvements after this.

I rebased changes from master here, and chose to use the FailureHandler that came from there. Please let me know if I did something wrong in this rebase.

After this gets merged there's some followup work to be done (when ctx logging hits GA): https://github.com/kubernetes/kubernetes/issues/111672

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Migrated kube-scheduler to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).
```


<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->

pohly committed:
scheduler_perf: support JSON

This enables benchmarking with JSON selected as output.

pohly committed:
scheduler: support contextual logging

This replaces all log calls through the global klog logger with a logger that
gets passed in by the caller, either through an explicit parameter or the
context. This makes it possible to produce more informative log output by
adding WithName and/or WithValues instrumentation (not done yet!). It also
makes "go test" failures more useful because the log output gets associated
with the test which produced it.

Usually, context is the more future-proof option. It has happened repeatedly
throughout the life of this commit that a function that originally had no
context later got changed in the master branch to accept for reasons unrelated
to contextual logging. In several places, passing a context can replace passing
a stop channel and then serves two purposes (access to logger and cancellation).

But klog.FromContext has some overhead, so for simple functions a logger is
passed.

pohly committed:
scheduler: use WithName and WithValues

This adds a prefix to each log entry that shows which operation and/or
component produced the log entry (WithName) and ensures that relevant
additional values, in particular the pod that gets scheduled, are always
included (WithValues).

This makes log entries easier to understand, in particular when multiple pods
get scheduled in parallel.

The downside is the constant overhead for these additional calls: they need to
do some work, whether log entries will be printed or not.

rikatz commented:
/ok-to-test

rikatz commented:
/sig scheduling

knelasevero commented:
/retest

knelasevero committed:
use klog's LoggerWithName/LoggerWithValues

knelasevero commented:
Noticed that in the in-between commit we were not using the LoggerWithValues/LoggerWithName wrapper functions that allow feature gate the contextualized logging. Had to throw that into all of pkg/scheduler/framework/runtime/framework.go

chendave reviewed (commented):
I guess this should hold until https://github.com/kubernetes/kubernetes/pull/111155 is merged? also, could you pls brief what's the log looks like before and after the change?

chendave commented:
> you mean https://github.com/kubernetes/kubernetes/pull/110833?

yes.

knelasevero commented:
> > you mean #110833?
> 
> yes.

No need to hold for that one. #111155 is a replacement for #110833

knelasevero commented:
/retest

leilajal commented:
/remove-sig api-machinery

ingvagabund commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -517,8 +515,8 @@ func (b *volumeBinder) bindAPIUpdate(ctx context.Context, pod *v1.Pod, bindings
 	// Update claims objects to trigger volume provisioning. Let the PV controller take care of the rest
 	// PV controller is expected to signal back by removing related annotations if actual provisioning fails
 	for i, claim = range claimsToProvision {
-		klog.V(5).InfoS("Updating claims objects to trigger volume provisioning", "pod", klog.KObj(pod), "PVC", klog.KObj(claim))
```
Why removing `pod:klog.KObj(pod)` pair? Is it already set in the logger some place else?

ingvagabund commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -332,13 +335,14 @@ func (pl *VolumeBinding) PreBind(ctx context.Context, cs *framework.CycleState,
+	logger.V(5).Info("Trying to bind volumes for pod")
+	err = pl.Binder.BindPodVolumes(logger, pod, podVolumes)
 	if err != nil {
-		klog.V(1).InfoS("Failed to bind volumes for pod", "pod", klog.KObj(pod), "err", err)
```
Why remove the pod?

ingvagabund commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -332,13 +335,14 @@ func (pl *VolumeBinding) PreBind(ctx context.Context, cs *framework.CycleState,
+		logger.V(1).Info("Failed to bind volumes for pod", "err", err)
 		return framework.AsStatus(err)
 	}
-	klog.V(5).InfoS("Success binding volumes for pod", "pod", klog.KObj(pod))
```
Why remove the pod?

ingvagabund commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -528,7 +543,7 @@ func fillEventToPluginMap(p framework.Plugin, eventToPlugins map[framework.Clust
 	// We treat it as: the plugin is not interested in any event, and hence pod failed by that plugin
 	// cannot be moved by any regular cluster event.
 	if len(events) == 0 {
-		klog.InfoS("Plugin's EventsToRegister() returned nil", "plugin", p.Name())
```
Why remove the plugin name key:value pair? Is the plugin name already set some place else?

ingvagabund commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -643,14 +658,15 @@ func (f *frameworkImpl) RunPreFilterExtensionAddPod(
 		if !status.IsSuccess() {
 			err := status.AsError()
-			klog.ErrorS(err, "Failed running AddPod on PreFilter plugin", "plugin", pl.Name(), "pod", klog.KObj(podToSchedule))
+			logger.Error(err, "Plugin failed")
```
Are really plugin, pod and possibly `AddPod` and `PreFilter` strings already in the logger context?

knelasevero replied:
Now they are, but still contextualized logging will be disabled by default while it is not GA, so added these keys where it was before, just changed them to be actually keys and not words inside the message. We need to remove these duplicate keys when  contextualized logging hits GA

ingvagabund commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -678,14 +694,15 @@ func (f *frameworkImpl) RunPreFilterExtensionRemovePod(
 		status = f.runPreFilterExtensionRemovePod(ctx, pl, state, podToSchedule, podInfoToRemove, nodeInfo)
 		if !status.IsSuccess() {
 			err := status.AsError()
-			klog.ErrorS(err, "Failed running RemovePod on PreFilter plugin", "plugin", pl.Name(), "pod", klog.KObj(podToSchedule))
```
Are really plugin, pod and possibly `RemovePod` and `PreFilter` strings already in the logger context? The same question for similar changes in other files below.

knelasevero replied:
https://github.com/kubernetes/kubernetes/pull/111155#discussion_r937049669

ingvagabund commented on pkg/scheduler/schedule_one.go:
```diff
@@ -312,8 +315,9 @@ func (sched *Scheduler) skipPodSchedule(fwk framework.Framework, pod *v1.Pod) bo
+	logger := klog.FromContext(ctx)
 
-	if err := sched.Cache.UpdateSnapshot(sched.nodeInfoSnapshot); err != nil {
+	if err := sched.Cache.UpdateSnapshot(logger, sched.nodeInfoSnapshot); err != nil {
```
`logger` used only once, `sched.Cache.UpdateSnapshot(klog.FromContext(ctx), sched.nodeInfoSnapshot)` will do as well

ingvagabund commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -613,7 +613,10 @@ func (f *frameworkImpl) RunPreFilterPlugins(ctx context.Context, state *framewor
 	}()
 	var result *framework.PreFilterResult
 	var pluginsWithNodes []string
+	logger := klog.FromContext(ctx).WithName("PreFilter").WithValues("pod", klog.KObj(pod))
```
Would be great to validate all the pod, nodes and other keys in the klog.Infos that got removed are properly replaced by `WithName` and `WithValues`. E.g. checking the logs and comparing the lines before and after.

pohly replied:
Right now, log lines should still include all extra values that are important, i.e. nothing should have been removed. That is because contextual logging is an optional feature and currently (because still alpha) disabled by default:

https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/3077-contextual-logging/README.md#redundant-keyvalue-pairs

Later when this is GA, the code can be simplified by removing redundant key/value pairs that are guaranteed to set via WithValues.

I agree that comparing logs would be useful, for both cases (contextual logging enabled and disabled). The problem will be that a simple diff will not work.

knelasevero replied:
I will compile and send here a comparison between 3 options:

- what we had before this PR
- What we have with this PR, but contextual logging enabled
- What we have with this PR, but contextual logging disabled

Would that be sufficient?

pohly replied:
SIG Scheduling is the first SIG to deal with this. The recommendation from the KEP is to keep the potentially duplicated key/value pairs. They will be filtered out by klog at runtime.

pohly commented on pkg/scheduler/apis/config/v1beta3/defaults.go:
```diff
@@ -101,6 +102,7 @@ func setDefaults_KubeSchedulerProfile(prof *v1beta3.KubeSchedulerProfile) {
 
 // SetDefaults_KubeSchedulerConfiguration sets additional defaults
 func SetDefaults_KubeSchedulerConfiguration(obj *v1beta3.KubeSchedulerConfiguration) {
+	logger := klog.TODO() // called by generated code which doesn't pass a logger
```
It's still a gap, though: if this code gets called in a unit test, it will log via the global klog logger and the log messages will not be visible in the output of a failed unit test.

I agree that this is probably not a big problem.

pohly commented on pkg/scheduler/framework/plugins/interpodaffinity/filtering.go:
```diff
@@ -255,7 +255,8 @@ func (pl *InterPodAffinity) PreFilter(ctx context.Context, cycleState *framework
 	}
-	s.namespaceLabels = GetNamespaceLabelsSnapshot(pod.Namespace, pl.nsLister)
+	logger := klog.FromContext(ctx)
+	s.namespaceLabels = GetNamespaceLabelsSnapshot(logger, pod.Namespace, pl.nsLister)
```
That's true, but then if later some other line gets added which also needs the logger, then this line with the `GetNamespaceLabelsSnapshot` call needs to be touched.

I prefer the separate FromContext call, but don't feel strongly about it.

pohly commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -775,11 +773,11 @@ func (b *volumeBinder) GetPodVolumes(pod *v1.Pod) (boundClaims []*v1.PersistentV
 	if err != nil {
 		// TODO: return the error once CSINode is created by default
-		klog.V(4).InfoS("Could not get a CSINode object for the node", "node", klog.KObj(node), "err", err)
+		logger.V(4).Info("Could not get a CSINode object for the node", "err", err)
```
See my [other comment](https://github.com/kubernetes/kubernetes/pull/111155#discussion_r935171071): right now, the node must be kept here for now.

This probably comes from a very early revision of my prototype PR where contextual logging was always enabled and the logger was guaranteed to have the node from a `WithValues` call. We don't have that guarantee at the moment.

knelasevero replied:
I grabbed the commit https://github.com/kubernetes/kubernetes/pull/110833/commits/01b8046c58e288a0a9d171235fb37d4abe0acc3e to go forward with this PR. 

Since it is not GA and since it can be disabled/enabled, how should we go forward with this? Add the node back and add a TODO to remove it once contextual logging is always enabled by default?

pohly replied:
Yes, add back all fields. A TODO at each line may be too much, but perhaps create a TODO in the place where you add the WithValues calls and a tracking issue as a reminder to revisit log calls once contextual logging is GA?

knelasevero replied:
done :heavy_check_mark:

knelasevero commented:
/remove-sig api-machinery

knelasevero commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -185,7 +185,7 @@ type SchedulerVolumeBinder interface {
 	//
 	// This function can be called in parallel.
-	BindPodVolumes(ctx context.Context, assumedPod *v1.Pod, podVolumes *PodVolumes) error
+	BindPodVolumes(logger klog.Logger, assumedPod *v1.Pod, podVolumes *PodVolumes) error
```
ctx came from rebases, in @pohly original PR ctx wasn't a argument, and logger was the new argument: https://github.com/pohly/kubernetes/blob/f7ac1e31be7fe27730b6a7dace2fdb35490e5f5c/pkg/scheduler/framework/plugins/volumebinding/binder.go

pohly replied:
If a function already takes a context, then keep that parameter and use the context to look up the logger.

knelasevero commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -500,15 +498,15 @@ func (b *volumeBinder) bindAPIUpdate(ctx context.Context, pod *v1.Pod, bindings
-		klog.V(2).InfoS("Claim bound to volume", "PVC", klog.KObj(binding.pvc), "PV", klog.KObj(binding.pv))
-		newPV, err := b.kubeClient.CoreV1().PersistentVolumes().Update(ctx, binding.pv, metav1.UpdateOptions{})
+		logger.V(2).Info("Claim bound to volume", "PVC", klog.KObj(binding.pvc), "PV", klog.KObj(binding.pv))
+		newPV, err := b.kubeClient.CoreV1().PersistentVolumes().Update(context.TODO(), binding.pv, metav1.UpdateOptions{})
```
See https://github.com/kubernetes/kubernetes/pull/111155#discussion_r936963083

@pohly @ingvagabund should I change all functions to only receive context or keep the logger?

knelasevero committed:
include duplicate k/v back, see #111672

knelasevero commented:
I went through places that I found and re-added duplicate key values that we want when contextualized logging is disabled and we still need those values. I also came back and passed pod/node objects where we used them for logging.

I added some comments in `pkg/scheduler/framework/runtime/framework.go` in places that we call WithValues so I don't add TODOs in every Info log entry.

I also created this issue: https://github.com/kubernetes/kubernetes/issues/111672. We can track this here while contextualized logging does not hit GA.

knelasevero commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -598,7 +613,21 @@ func (f *frameworkImpl) RunPreFilterPlugins(ctx context.Context, state *framewor
 	var pluginsWithNodes []string
+	logger := klog.FromContext(ctx)
+	logger = klog.LoggerWithName(logger, "PreFilter")
+	// We are adding LoggerWithValues for contextualized logging here
+	// but contextualized logging is feature gated and optional until
+	// it hits GA when it will be defaulted. For that reason
+	// logger.V(<LOGLEVEL>).Info or Error calls still need to include
+	// WithValues Key/Values while contextualized logging does not hit GA
+	// (duplicates get filtered out if contextualized logging is enabled)
+	// TODO(knelasevero): Remove duplicated keys from log entry calls
+	// When contextualized logging hits GA
+	// https://github.com/kubernetes/kubernetes/issues/111672
```
Added these comments in places where we call LoggerWithValues

knelasevero commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -643,14 +672,28 @@ func (f *frameworkImpl) RunPreFilterExtensionAddPod(
 		if !status.IsSuccess() {
 			err := status.AsError()
-			klog.ErrorS(err, "Failed running AddPod on PreFilter plugin", "plugin", pl.Name(), "pod", klog.KObj(podToSchedule))
+			logger.Error(err, "Plugin failed", "pod", klog.KObj(podToSchedule), "node", klog.KObj(nodeInfo.Node()), "extensibleAPI", "PreFilter", "extension", "AddPod", "plugin", pl.Name())
```
I am not sure about this nomenclature. Extensible APIs is something that I saw being used in some of the scheduler framework documentations for `PreFilter`/`Filter`/etc. I am calling `AddPod`/`RemovePod`/etc extensions here .... I am open to better suggestions.

knelasevero commented on pkg/scheduler/apis/config/v1beta2/default_plugins.go:
```diff
@@ -107,34 +107,34 @@ func getDefaultPlugins() *v1beta2.Plugins {
 }
 
-func applyFeatureGates(config *v1beta2.Plugins) {
+func applyFeatureGates(logger klog.Logger, config *v1beta2.Plugins) {
```
Got this from @pohly initial work. @pohly is the idea here standardization?

pohly replied:
It probably was used at some point. If that is not the case anymore, then there's no need to add it now because this is an internal function that can still be changed later if (and when) needed.

alculquicondor replied:
v1beta2 and v1beta3 packages are already marked as deprecated. No need to add now.

cici37 commented:
/remove-sig api-machinery

knelasevero commented:
https://gist.github.com/knelasevero/ce0b25bec68e42dc5b9f0ed88baf09e9#file-normal-log-log

https://gist.github.com/knelasevero/ce0b25bec68e42dc5b9f0ed88baf09e9#file-ctx-log-log


Here I have a comparison between logs being spit by **kube-scheduler released on 1.24** and **kube-scheeduler with contextualized logging** with code changes and flag `-feature-gates="ContextualLogging=true"`

It wont be that different to be honest. As per https://github.com/kubernetes/kubernetes/pull/111155#discussion_r935961069 this PR was initially not adding those keys in the log entries, and relying on WithName/WithValues calls. However later we decided re-add those keys while  ContextualLogging is not enabled by default. Later, when this hits GA, we need to revisit this and work on this issue: https://github.com/kubernetes/kubernetes/issues/111672. 

What is happening right now: 

- With `-feature-gates="ContextualLogging=true"` klog is seeing some keys being added directly in the log entry calls, and some being added in WithValues/WithName calls, and it is filtering out duplicated ones. 
- With the code from the released 1.24 kube-scheduler we are just using keys being added directly into klog log entries.

cc @pohly @ingvagabund

knelasevero committed:
logcheck and fmt

knelasevero commented:
/remove-sig api-machinery

fedebongio commented:
/remove-sig api-machinery

knelasevero commented:
hey @dims 

we were discussing a few things before coming back to this pr: https://github.com/kubernetes/kubernetes/issues/91633#issuecomment-1346587315

I will rebase it when I am back from vacation after NYE

sftim commented:
The changelog entry I suggest for this:
```markdown
Migrated kube-scheduler to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).
```

knelasevero committed:
convert missing entries to ctx logging

knelasevero commented:
/remove-sig api-machinery

knelasevero commented:
```
pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:548:92: function "KObjSlice" should not be used, convert to contextual logging (logcheck)
                logger.V(5).Info("remembering potential nodes", "pod", klog.KObj(pod), "potentialnodes", klog.KObjSlice(nodes))
```

Hey @pohly what should be used instead of klog.KObjSlice so we don't have ctx-logging warnings here?

knelasevero commented:
/remove-sig api-machinery

pohly commented:
```
running golangci-lint 
pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:548:92: function "KObjSlice" should not be used, convert to contextual logging (logcheck)
		logger.V(5).Info("remembering potential nodes", "pod", klog.KObj(pod), "potentialnodes", klog.KObjSlice(nodes))
		                                                                                         ^
```

This is a false positive. `klog.KObjSlice` should be usable without such an error. This was already fixed, but Kubernetes wasn't using the latest logcheck -> https://github.com/kubernetes/kubernetes/pull/115060

knelasevero commented:
Thanks!

knelasevero commented:
/open

knelasevero commented:
I will be rebasing this soon

knelasevero commented:
Funny thing that Pull Requests are basically issues under the hood for github, @pohly . So when you wrote fixes #pr/comment, it actually closed the PR :joy:

knelasevero commented:
/remove-sig api-machinery

knelasevero commented:
/assign @huang-wei @mikedanese @alculquicondor

Now that https://github.com/kubernetes/kubernetes/pull/115060 is merged we can go forward here.

Hey @alculquicondor so, can we review/merge this one as per https://github.com/kubernetes/kubernetes/issues/91633#issuecomment-1346574986

> @knelasevero @pohly we are early in the 1.27 release cycle. I think we should merge what we have and monitor the performance dashboard. If there are issues, we try to optimize in follow up PRs or (worst case) revert.

I would like to point back to some stuff here in the PR since they are lost up in the bots update messages:

https://github.com/kubernetes/kubernetes/pull/111155#discussion_r937044514

I said I found this in the scheduling framework docs, but now it seems its changed. Please let me know if this should be called something else.

knelasevero commented:
/remove-sig api-machinery

alculquicondor commented:
why are there changes to kube-controller-manager in the same PR?
Can you split any change not related to the scheduler into a separate PR? If we need some refactoring before being able to change scheduler, please also split that into a separate PR.

pohly commented:
Also beware that https://github.com/kubernetes/kubernetes/pull/114843#pullrequestreview-1267621654 includes the changes for contextual logging under test/integration. It would be best to merge that first and then rebase this one here.

knelasevero commented:
@alculquicondor @pohly : would this be acceptable as the separate PR?

https://github.com/kubernetes/kubernetes/pull/115298

knelasevero commented:
/label tide/merge-method-squash

knelasevero commented:
Should we wait for https://github.com/kubernetes/kubernetes/pull/114843#pullrequestreview-1267621654

before going forward here then? @pohly

alculquicondor commented:
It's in my review queue. Let me have a quick look.

knelasevero commented:
/test pull-kubernetes-node-e2e-containerd

alculquicondor commented:
Please rebase

knelasevero commented:
Hey hey :wave: , this one is ready for review. Rebased already

alculquicondor commented on cmd/kube-scheduler/app/options/options_test.go:
```diff
@@ -1662,7 +1662,7 @@ profiles:
 			}
 			// create the config
-			config, err := tc.options.Config()
+			config, err := tc.options.Config(context.Background())
```
Use `ktesting.NewTestContext(t)`

alculquicondor commented on cmd/kube-scheduler/app/server.go:
```diff
@@ -111,6 +111,7 @@ for more information about scheduling and the kube-scheduler component.`,
 
 	if err := cmd.MarkFlagFilename("config", "yaml", "yml", "json"); err != nil {
+		// nolint:logcheck // TODO (?): return error instead of logging it here
 		klog.ErrorS(err, "Failed to mark flag filename")
```
can you just use the global logger?

pohly replied:
Logging shouldn't be used during program init because it hasn't been configured yet, so returning an error and then dealing with that as part of the "normal" program output is better. Having said that, we can also decide that it doesn't matter because this line will never be reached.

Then both nolint or `klog.Background()` are okay. I prefer to decide about the TODO now and lean towards `klog.Background`.

alculquicondor commented on hack/logcheck.conf:
```diff
@@ -15,18 +15,18 @@
 # The following packages have been migrated to contextual logging.
 # Packages matched here do not have to be listed above because
 # "contextual" implies "structured".
 # TODO next: contextual k8s.io/kubernetes/pkg/scheduler/.*
```
remove this line

alculquicondor commented on pkg/scheduler/apis/config/v1beta2/defaults_test.go:
```diff
@@ -150,7 +152,7 @@ func TestSchedulerDefaults(t *testing.T) {
 				Profiles: []v1beta2.KubeSchedulerProfile{
 					{
-						Plugins:       getDefaultPlugins(),
+						Plugins:       getDefaultPlugins(logger),
```
revert

alculquicondor commented on pkg/scheduler/apis/config/v1beta3/default_plugins.go:
```diff
@@ -27,7 +27,7 @@ import (
 
 // getDefaultPlugins returns the default set of plugins.
-func getDefaultPlugins() *v1beta3.Plugins {
+func getDefaultPlugins(logger klog.Logger) *v1beta3.Plugins {
```
revert

alculquicondor commented on pkg/scheduler/apis/config/v1beta3/defaults_test.go:
```diff
@@ -150,7 +152,7 @@ func TestSchedulerDefaults(t *testing.T) {
 				Profiles: []v1beta3.KubeSchedulerProfile{
 					{
-						Plugins:       getDefaultPlugins(),
+						Plugins:       getDefaultPlugins(logger),
```
revert

alculquicondor commented on pkg/scheduler/apis/config/v1beta3/defaults_test.go:
```diff
@@ -187,8 +189,8 @@ func TestSchedulerDefaults(t *testing.T) {
 					{
-						SchedulerName: pointer.String("default-scheduler"),
-						Plugins:       getDefaultPlugins(),
+						SchedulerName: pointer.StringPtr("default-scheduler"),
```
revert this line

pohly commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -56,42 +57,45 @@ func (sched *Scheduler) onStorageClassAdd(obj interface{}) {
 }
 
 func (sched *Scheduler) addNodeToCache(obj interface{}) {
+	logger := sched.logger
```
@alculquicondor asked me about this.

Using a per-object logger instance is okay here because these callbacks are not getting a context from their caller.

But the comment for sched.logger should be very explicit that it is only for such background activities and that it is not to be used when the callchain has a logger.

pohly commented on pkg/scheduler/apis/config/v1/defaults.go:
```diff
@@ -102,6 +103,7 @@ func setDefaults_KubeSchedulerProfile(prof *configv1.KubeSchedulerProfile) {
 
 // SetDefaults_KubeSchedulerConfiguration sets additional defaults
 func SetDefaults_KubeSchedulerConfiguration(obj *configv1.KubeSchedulerConfiguration) {
+	logger := klog.TODO() // called by generated code that doesn't pass a logger
```
Same comment as in https://github.com/kubernetes/kubernetes/pull/111155/files#r929904171: it would be nicer if the defaulting code passed down a context, but at this point it's not a big problem and thus not required.

alculquicondor replied:
so should we use `context.Background` instead?

pohly replied:
I prefer `klog.TODO` because ultimately this is something that should get fixed - it's just not important. `klog.Background` implies that the code is fine.

Compare with `ctx.TODO` after client-go started to require it: the Kubernetes code base is still littered with it. At some point, someone needs to find a solution for each of those calls. Using `ctx.Background` would hide them.

alculquicondor reviewed (commented):
First pass, mostly on files that are only involved in startup, but not in scheduling cycles.

With that said, I wonder if this PR can be split in two. For example, the first PR could leave out everything in eventhandlers.go, schedule_one.go and pkg/scheduler/framework, pkg/scheduler/internal, which require more thorough reviews.

knelasevero commented:
> First pass, mostly on files that are only involved in startup, but not in scheduling cycles.
> 
> With that said, I wonder if this PR can be split in two. For example, the first PR could leave out everything in eventhandlers.go, schedule_one.go and pkg/scheduler/framework, pkg/scheduler/internal, which require more thorough reviews.

I've moved the changes that are not related to those files to this PR: https://github.com/kubernetes/kubernetes/pull/115588

I prefer to keep the bulk of the internal changes here, since most of the discussion about it is here. I already included the changes you requested there. I also reverted hack/logcheck changes and added more comments since we cant let logcheck complain until we migrate everything

knelasevero commented:
I won't be rebasing this one yet, as it will make more sense to rebase after https://github.com/kubernetes/kubernetes/pull/115588 gets merged

knelasevero commented:
@alculquicondor @pohly  rebased. I may have missed somethings because there were quite a lot of conflicts. I will check more thoroughly tomorrow, but feel free to have some pass troughs if you have time

knelasevero committed:
revert scheduler_perf changes

knelasevero commented:
/test pull-kubernetes-e2e-gce-100-performance

knelasevero commented:
/retest

alculquicondor commented:
rebase so that the changes already merged don't show up here #115588

knelasevero commented:
hmm :thinking: I rebased after that. I will check tomorrow why they are showing

knelasevero commented:
@alculquicondor I don't see those changes here. The only change in cmd/ is because of changes in the scheduler.New inside of pkg/

alculquicondor commented:
ah, I was confused by the changes in `pkg/scheduler/apis/config/v1/default_plugins.go`. I thought they had all merged in the previous PR.

alculquicondor commented:
any chance you can split the PRs again (and maybe a few others too)? I have very limited time to review and maybe we won't be able to merge this entire PR, but we can still make progress.

knelasevero commented:
Maybe someone from the community would want to take that task and push these splits of this PR forward? Since they can use this PR as reference it would be easy to anyone wanting to contribute (I can help, of course, but I may have to focus on some other things).

alculquicondor commented:
/help

knelasevero commented:
/help-wanted

alculquicondor commented:
Maybe it doesn't work on PRs

pohly commented:
@alculquicondor: any suggestion on how to split this up? By directories?

Looking at the commit history, all commits probably can be squashed into one, i.e. there are no commits which could be merged separately.

alculquicondor commented:
Everything under `pkg/scheduler/apis` can be its own PR.

The rest is a bit unclear. Maybe it's possible to have a PR for each of the folders in `pkg/scheduler/framework/plugins`. Followed by a PR that does the rest.

ncdc commented:
I might be able to do some, but I wouldn't be able to start on anything for at least 2 weeks

alculquicondor commented:
code freeze is tomorrow, so I think the boat has sailed for this release.

Namanl2001 commented:
I would like to pick this up, but unfortunately, this won't be possible to be done in a couple of days

mengjiao-liu commented:
I can help too, but time is too urgent before the code freeze, both for contributors and reviewers, maybe we can finish it in the next version.

mengjiao-liu commented:
> If we start the task, we need to figure out how to split PR.
I am not quite clear about the directory structure of the scheduler, so I will first divide PR from the direction provided by @alculquicondor and then adjust it according to the actual implementation.

Please see  https://github.com/kubernetes/kubernetes/issues/91633#issuecomment-1467645065.

mengjiao-liu commented:
/sig instrumentation
/wg structured-logging
/area logging

mengjiao-liu commented:
Work continues in https://github.com/kubernetes/kubernetes/issues/91633#issuecomment-1467645065

/close
